### PR TITLE
Exposure for indoors

### PR DESCRIPTION
**Pull request by jgoppert**
@LorenzMeier @bkueng @ecmnet @mhkabir please test, this is a huge improvement for indoors and I suspect outdoors as well.
### Changes
- Add a lot more control over camera gain and exposure, can set with params not to allow fast tuning
- Tune exposure and gain settings to work well with low light and outdoor/ bright light settings using auto gain and exposure control.
- Instead of publishing flow every 10 received frames, publishes at fixed rate set by PUB_RATE param

### CODE REVIEW

**Comment by ecmnet** [comment]
@jgoppert I've not setup my development environment for PX4Flow. Could you provide a binary to upload?

**Comment by jgoppert** [comment]
@ecmnet https://drive.google.com/open?id=0B3wIEs1gcVahNktlQkhCMndIeUE

**Comment by ecmnet** [comment]
@jgoppert Thanks a lot! I'll give it a try a soon as I've mounted flow to my device again.

**Comment by ecmnet** [comment]
@jgoppert Unfortunately, my PX4Flow does not work after upgrading (does not connect to QGC at all)

**Comment by jgoppert** [comment]
Weird, will try uploading the image I created for you tonight on my device.

**Comment by ecmnet** [comment]
@jgoppert I've setup my environment now and was able to compile and upload this branch. Is it correct to have some white square dots in the image now?

**Comment by jgoppert** [comment]
Yes, the dots indicate high quality optical flow in that region of the image.

**Comment by jgoppert** [comment]
Or more specifically,  that 8x8 tile had enough difference between its pixels (f thrsh param) that it included that tile in the optical flow calculation.

**Comment by ecmnet** [comment]
@jgoppert Thanks for clarification. I will do some testflights as soon I've mounted flow again.

**Comment by ecmnet** [comment]
@jgoppert Flow quality seems not to be reasonable with this patch. Even, if I darken the lens, quality is reported at 220-250 in my case.

**Comment by jgoppert** [comment]
It increases the gain and exposure when you cover the lens so you start picking up noise.  Look at the image in qgc. The max gain and max exposure can be lowered to limit the effect, but if you don't cover the lens I think it gives good low and high light performance.

File: src/modules/flow/main.c [hunk 1]
```diff
@@ -699,7 +707,7 @@ int main(void)
 			uint16_t frame = 0;
 			for (frame = 0; frame < image_size_send / MAVLINK_MSG_ENCAPSULATED_DATA_FIELD_DATA_LEN + 1; frame++)
 			{
-				mavlink_msg_encapsulated_data_send(MAVLINK_COMM_2, frame, &((uint8_t *) current_image)[frame * MAVLINK_MSG_ENCAPSULATED_DATA_FIELD_DATA_LEN]);
+				mavlink_msg_encapsulated_data_send(MAVLINK_COMM_2, frame, &((uint8_t *) previous_image)[frame * MAVLINK_MSG_ENCAPSULATED_DATA_FIELD_DATA_LEN]);
```
Comment (ChristophTobler) [inline src/modules/flow/main.c]: Why did you change this to `previous_image`?

File: src/modules/flow/mt9v034.c [hunk 2]
```diff
@@ -79,7 +79,7 @@ void mt9v034_context_configuration(void)
 	uint16_t new_height_context_b = FULL_IMAGE_COLUMN_SIZE * 4;
 
 	/* blanking settings */
-	uint16_t new_hor_blanking_context_a = 709 + MINIMUM_HORIZONTAL_BLANKING;// 709 is minimum value without distortions
+	uint16_t new_hor_blanking_context_a = 425 + MINIMUM_HORIZONTAL_BLANKING;// 709 is minimum value without distortions
```
Comment (ChristophTobler) [inline src/modules/flow/mt9v034.c]: comment needs to be adapted

**Comment by lericson** [comment]
I've tested this and it doesn't work very well at all for me. It might be my hardware of course but I get pretty heavy banding with this.

**Comment by jgoppert** [comment]
@lericson Go into qgc and play with exposure_max and brightness. Lowering brightness down should fix it for you (try brightness 10, max_exposure 300). I've tested this a lot, so trust me, it is a huge difference for indoors. But watch out for specular reflections and shadows.

**Comment by lericson** [comment]
@jgoppert I played around a bit with the parameters. Couldn't get very good results sadly. For your perusing, here is my merged version of this PR and #90: https://github.com/KTH-CAS-UAV/Flow/tree/master-with-exposure-pr

**Comment by jgoppert** [comment]
Don't have time to load it myself. What does your flow quality say? It should be much higher with this patch. Also, when you evaluate the quality, don't just look at the image, enable the flat SEND_LPOS. This will give you output of a system onboard estimator of the position, assuming that the board is always aligned with north.

File: src/modules/flow/main.c (hunk 1 above)
Comment (jgoppert) [inline src/modules/flow/main.c]: so that it gets overlayed with the dots for flow quality

**Comment by lericson** [comment]
It said 255 all the time and the image quality was not very good with heavy banding. The flow estimation was also very noisy, which isn't so surprising. 🙂

- Ludvig

**Comment by jgoppert** [comment]
When you set the parameters it doesn't stick after a reboot, if you set brightness to 8 and max exposure to 200 (in the firmware and rebulid), that is bascially the same as the default firmware. This firmware is basically just giving you more control over the settings.

File: makefiles/baremetal/toolchain_gnu-arm-eabi.mk [hunk 3]
```diff
@@ -50,7 +50,7 @@ OBJDUMP			 = $(CROSSDEV)objdump
 
 # Check if the right version of the toolchain is available
 #
-CROSSDEV_VER_SUPPORTED	 = 4.7.4 4.7.5 4.7.6 4.8.4 4.9.3
+CROSSDEV_VER_SUPPORTED	 = 4.7.4 4.7.5 4.7.6 4.8.4 4.9.3 5.4.1
```
Comment (jgoppert) [inline makefiles/baremetal/toolchain_gnu-arm-eabi.mk]: This supports the latest gcc, so I included it here.

File: src/include/settings.h [hunk 4]
```diff
@@ -40,7 +40,7 @@
 #define ONBOARD_PARAM_NAME_LENGTH 		15
 #define BOTTOM_FLOW_IMAGE_HEIGHT		64
 #define BOTTOM_FLOW_IMAGE_WIDTH			64
-#define BOTTOM_FLOW_SEARCH_WINDOW_SIZE 	4
+#define BOTTOM_FLOW_SEARCH_WINDOW_SIZE 	8
```
Comment (jgoppert) [inline src/include/settings.h]: You can play with this, but this makes it possible for it to see a large max velocity, but you have to watch the update rate and make sure the cpu isn't maxing out.

File: src/include/settings.h [hunk 5]
```diff
@@ -139,9 +139,18 @@ enum global_param_id_t
...
 	PARAM_SENSOR_POSITION,
+	PARAM_GAIN_MAX,
+	PARAM_EXPOSURE_MAX,
+	PARAM_SHTR_W_1,
+	PARAM_SHTR_W_2,
+	PARAM_SHTR_W_TOT,
+	PARAM_HDR,
+	PARAM_AEC,
+	PARAM_AGC,
+	PARAM_BRIGHT,
```
Comment (jgoppert) [inline src/include/settings.h]: Here I exposed settings that were hard coded in the firmware.

File: src/modules/flow/communication.c [hunk 6]
```diff
@@ -273,6 +273,23 @@ void handle_mavlink_message(mavlink_channel_t chan,
...
+									i == PARAM_AGC ||
+									i == PARAM_BRIGHT ||
+									i == PARAM_GAIN_MAX
+									)
+							{
+								mt9v034_context_configuration();
+								dma_reconfigure();
+								buffer_reset();
+							}
+
```
Comment (jgoppert) [inline src/modules/flow/communication.c]: This is just setting the parameters that you request in QGC.

File: src/modules/flow/flow.c [hunk 7]
```diff
@@ -501,23 +501,23 @@ uint8_t compute_flow(uint8_t *image1, uint8_t *image2, float x_rate, float y_rat
...
+
+				uint32_t diff = compute_diff(image1, i, j, (uint16_t) global_data.param[PARAM_IMAGE_WIDTH]);
+				if (diff > global_data.param[PARAM_BOTTOM_FLOW_FEATURE_THRESHOLD])
+				{
+					image1[j * ((uint16_t) global_data.param[PARAM_IMAGE_WIDTH]) + i] = 255;
+				}
+
+			}
+		}
+	}
```
Comment (jgoppert) [inline src/modules/flow/flow.c]: This piece of code was commented out and shows you flow quality indicators for each grid cell.

File: src/modules/flow/main.c [hunk 8]
```diff
@@ -533,8 +539,11 @@ int main(void)
                         PROBE_3(true);
 
             //serial mavlink  + usb mavlink output throttled
-			if (counter % (uint32_t)global_data.param[PARAM_BOTTOM_FLOW_SERIAL_THROTTLE_FACTOR] == 0)//throttling factor
+			uint32_t now = get_boot_time_us();
+			uint32_t time_since_last_pub = now - time_last_pub;
+			if (time_since_last_pub > (1.0e6f/global_data.param[PARAM_BOTTOM_FLOW_PUB_RATE]))
```
Comment (jgoppert) [inline src/modules/flow/main.c]: This is the logic to publish at a fixed rate instead of based on the exposure setting.

File: src/modules/flow/main.c [hunk 9]
```diff
@@ -577,10 +586,9 @@ int main(void)
 					lpos.x += ground_distance*accumulated_flow_x;
 					lpos.y += ground_distance*accumulated_flow_y;
 					lpos.z = -ground_distance;
- 					/* velocity not directly measured and not important for testing */
-					lpos.vx = 0;
-					lpos.vy = 0;
-					lpos.vz = 0;
+					lpos.vx = ground_distance*accumulated_flow_x/integration_timespan;
+					lpos.vy = ground_distance*accumulated_flow_y/integration_timespan;
+					lpos.vz = 0; // no direct measurement, just ignore
```
Comment (jgoppert) [inline src/modules/flow/main.c]: I included the velocity here for debug.

File: src/modules/flow/settings.c [hunk 10]
```diff
@@ -195,9 +195,45 @@ void global_data_reset_param_defaults(void){
...
+
+	global_data.param[PARAM_AEC] = 1;
+	strcpy(global_data.param_name[PARAM_AEC], "AEC");
+	global_data.param_access[PARAM_AEC] = READ_WRITE;
+
+	global_data.param[PARAM_AGC] = 1;
+	strcpy(global_data.param_name[PARAM_AGC], "AGC");
+	global_data.param_access[PARAM_AGC] = READ_WRITE;
+
+	global_data.param[PARAM_BRIGHT] = 20;
```
Comment (jgoppert) [inline src/modules/flow/settings.c]: If you have trouble with washout/ banding, try lowering this value. 8 is the minimum.

File: src/modules/flow/settings.c [hunk 11]
```diff
@@ -195,9 +195,45 @@ void global_data_reset_param_defaults(void){
...
 	global_data.param_access[PARAM_BOTTOM_FLOW_WEIGHT_NEW] = READ_WRITE;
 
-	global_data.param[PARAM_BOTTOM_FLOW_SERIAL_THROTTLE_FACTOR] = 10.0f;
-	strcpy(global_data.param_name[PARAM_BOTTOM_FLOW_SERIAL_THROTTLE_FACTOR], "BFLOW_THROTT");
-	global_data.param_access[PARAM_BOTTOM_FLOW_SERIAL_THROTTLE_FACTOR] = READ_WRITE;
+	global_data.param[PARAM_BOTTOM_FLOW_PUB_RATE] = 10.0f;
+	strcpy(global_data.param_name[PARAM_BOTTOM_FLOW_PUB_RATE], "BFLOW_RATE");
+	global_data.param_access[PARAM_BOTTOM_FLOW_PUB_RATE] = READ_WRITE;
+
+	global_data.param[PARAM_EXPOSURE_MAX] = 500;
```
Comment (jgoppert) [inline src/modules/flow/settings.c]: Increasing this value helps in low light, but it also makes the max velocity you can go lower. If you have trouble with velocity maxing out, lower this to around 200 or 100.

File: src/modules/flow/settings.c [hunk 12]
```diff
@@ -195,9 +195,45 @@ void global_data_reset_param_defaults(void){
...
+	strcpy(global_data.param_name[PARAM_EXPOSURE_MAX], "EXPOSURE_MAX");
+	global_data.param_access[PARAM_EXPOSURE_MAX] = READ_WRITE;
+
+	global_data.param[PARAM_GAIN_MAX] = 16;
```
Comment (jgoppert) [inline src/modules/flow/settings.c]: Gain can make the image pixelated. If your image has a lot of noise, try lowering this to 4 or 8.

**Comment by jgoppert** [comment]
@lericson  I tried to explain my changes in this patch, hope that helps and you find some of it useful for your work.

**Comment by ChristophTobler** [comment]
@jgoppert I have bench tested this PR so far. Looks very good. I increased the flow feature threshold because it reported good quality even though it was just noisy and pointing to a plain white surface. I think it's safer this way.
